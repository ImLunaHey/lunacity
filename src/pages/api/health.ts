import type { NextApiRequest, NextApiResponse } from 'next'

const Health = (_request: NextApiRequest, response: NextApiResponse) => {
    response.status(200);
};

export default Health;
