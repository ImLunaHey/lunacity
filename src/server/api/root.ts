import { createTRPCRouter } from './trpc';
import { pageRouter } from './routers/page';
import { tagRouter } from './routers/tag';

/**
 * This is the primary router for your server.
 *
 * All routers added in /api/routers should be manually added here
 */
export const appRouter = createTRPCRouter({
  page: pageRouter,
  tag: tagRouter,
});

// export type definition of API
export type AppRouter = typeof appRouter;
