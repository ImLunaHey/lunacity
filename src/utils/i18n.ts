import i18nNext, { type InitOptions } from 'i18next';
import { initReactI18next } from 'react-i18next';

import Backend, { type HttpBackendOptions } from 'i18next-http-backend';
import LanguageDetector from 'i18next-browser-languagedetector';
import getConfig from 'next/config';
import Pseudo from 'i18next-pseudo';
import { env } from '@app/env.mjs';

const { publicRuntimeConfig } = getConfig() as { publicRuntimeConfig: { APP_URL: string; WS_URL: string; } };
const { APP_URL } = publicRuntimeConfig;

const options = {
    fallbackLng: 'en',
    supportedLngs: ['en', 'ar', 'cn', 'es', 'pt'],
    debug: !!env.NEXT_PUBLIC_LANG_DEBUG,
    defaultNS: 'common',
    returnNull: false,
    load: 'languageOnly' as const,
    react: {
        useSuspense: false,
    },
    ns: ['common'],
    interpolation: {
        escapeValue: false, // not needed for react as it escapes by default
    },
    nonExplicitSupportedLngs: true,
    backend: {
        loadPath: `${APP_URL}/locales/{{lng}}/{{ns}}.json`,
    },
    detection: {
        caches: []
    },
    postProcess: ['pseudo'],
} satisfies InitOptions<HttpBackendOptions>;

export const i18n = i18nNext
    .use(new Pseudo({
        enabled: env.NEXT_PUBLIC_PSUDO_ENABLED === 'true',
        languageToPseudo: 'en-US',
    }))
    // load translation using http -> see /public/locales (i.e. https://github.com/i18next/react-i18next/tree/master/example/react/public/locales)
    // learn more: https://github.com/i18next/i18next-http-backend
    .use(Backend)
    // detect user language
    // learn more: https://github.com/i18next/i18next-browser-languageDetector
    .use(LanguageDetector)
    // pass the i18n instance to react-i18next.
    .use(initReactI18next)
    // init i18next
    // for all options read: https://www.i18next.com/overview/configuration-options
    .init<HttpBackendOptions>(options);
