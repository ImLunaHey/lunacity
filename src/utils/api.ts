/**
 * This is the client-side entrypoint for your tRPC API.
 * It's used to create the `api` object which contains the Next.js App-wrapper
 * as well as your typesafe react-query hooks.
 *
 * We also create a few inference helpers for input and output types
 */
import { createWSClient, httpBatchLink, loggerLink, splitLink, wsLink } from '@trpc/client';
import { createTRPCNext } from '@trpc/next';
import { type inferRouterInputs, type inferRouterOutputs } from '@trpc/server';
import { type NextPageContext } from 'next';
import getConfig from 'next/config';
import superJSON from 'superjson';
import { type AppRouter } from '@app/server/api/root';

const { publicRuntimeConfig } = getConfig() as { publicRuntimeConfig: { APP_URL: string; WS_URL: string; } };
const { APP_URL, WS_URL } = publicRuntimeConfig;
const getEndingLink = (ctx: NextPageContext | undefined) => {
  if (typeof window === 'undefined') {
    return httpBatchLink({
      url: `${APP_URL}/api/trpc`,
      headers() {
        if (ctx?.req) {
          // on ssr, forward client's headers to the server
          return {
            ...ctx.req.headers,
            'x-ssr': '1',
          };
        }
        return {};
      },
    });
  }

  // Make Websocket client
  const client = createWSClient({
    url: WS_URL,
  });

  // If we're on the client create a split link that sends both queries
  // and mutations over http-batch and all subscriptions over websocket
  return splitLink({
    condition(op) {
      return op.type !== 'subscription';
    },
    // HTTP batching
    true: httpBatchLink({
      url: `${APP_URL}/api/trpc`,
      headers() {
        if (ctx?.req) {
          // on ssr, forward client's headers to the server
          return {
            ...ctx.req.headers,
            'x-ssr': '1',
          };
        }
        return {};
      },
    }),
    // Websocket
    false: wsLink<AppRouter>({
      client,
    })
  });
}

/**
 * A set of typesafe react-query hooks for your tRPC API
 */
export const api = createTRPCNext<AppRouter>({
  config({ ctx }) {
    return {
      /**
       * Transformer used for data de-serialization from the server
       * @see https://trpc.io/docs/data-transformers
       **/
      transformer: superJSON,

      /**
       * Links used to determine request flow from client to server
       * @see https://trpc.io/docs/links
       * */
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === 'development' ||
            (opts.direction === 'down' && opts.result instanceof Error),
        }),
        getEndingLink(ctx)
      ]
    };
  },
  /**
   * Whether tRPC should await queries when server rendering pages
   * @see https://trpc.io/docs/nextjs#ssr-boolean-default-false
   */
  ssr: false,
});

/**
 * Inference helper for inputs
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;
/**
 * Inference helper for outputs
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
