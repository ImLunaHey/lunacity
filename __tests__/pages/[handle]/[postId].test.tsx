import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import PostId, { getServerSideProps } from '@app/pages/[handle]/[postId]';
import createPrismaMock from 'prisma-mock';
import { randomUUID } from 'crypto';
import { createMockUser } from '__tests__/__utils__/mocks/create-mock-user';
import { createMockSession } from '__tests__/__utils__/mocks/create-mock-session';
import { api } from '@app/utils/api';
import type { IncomingMessage, ServerResponse } from 'http';

// Mock next-auth with defaults
jest.mock('next/config', () => ({
  __esModule: true,
  default: () => ({
    publicRuntimeConfig: {
      APP_URL: 'http://localhost:3000',
      WS_URL: 'ws://localhost:3001',
    },
  }),
}));

// Mock text-post component
jest.mock('@app/components/text-post', () => ({
  __esModule: true,
  TextPost: () => <div>__TEXT_POST__MOCK__</div>,
}));

jest.mock('react-i18next', () => ({
  useTranslation: jest.fn(() => ({
    t: jest.fn((text: string, args) => (args ? `${text} ${JSON.stringify(args, null)}` : text)),
    ready: true,
  })),
}));

jest.mock('@app/utils/api', () => ({
  api: {
    post: {
      getPostDetails: {
        useQuery: jest.fn(() => ({
          isLoading: true,
          data: undefined,
        })),
      },
    },
  },
}));

jest.mock('@app/common/with-public-access', () => ({
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-explicit-any
  withPublicAccess: (fn: any) => (ctx: any) => fn(ctx),
}));

describe('getServerSideProps', () => {
  it('returns the postId from the url', async () => {
    const postId = '1';
    const result = await getServerSideProps({
      query: {
        postId,
      },
      req: {} as IncomingMessage & { cookies: Partial<{ [key: string]: string }> },
      res: {} as ServerResponse<IncomingMessage>,
      resolvedUrl: '/',
    });
    expect(result).toEqual({
      props: {
        postId,
      },
    });
  });
});

describe('PostId', () => {
  it('throws an error when no postId is provided', () => {
    // @ts-expect-error - Mocking useQuery
    jest.mocked(api.post.getPostDetails.useQuery).mockImplementation(() => ({
      isLoading: false,
      data: null,
    }));

    const { container } = render(<PostId postId="" />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('h1')).toHaveTextContent('404');
    expect(container.querySelector('h2')).toHaveTextContent('errors.post.not-found.');
  });

  it('renders a loading state', () => {
    // @ts-expect-error - Mocking useQuery
    jest.mocked(api.post.getPostDetails.useQuery).mockImplementation(() => ({
      isLoading: true,
      data: null,
    }));

    const { container } = render(<PostId postId="1" />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('span')?.querySelector('span')).toHaveAttribute('aria-label', 'Loading');
  });

  it('renders a post with the provided ID', async () => {
    const session = createMockSession();
    const prismaMock = createPrismaMock<NonNullable<typeof prisma>>();
    const post = await prismaMock.post.create({
      data: {
        id: '1',
        title: 'Test Post',
        body: 'This is a test post',
        createdAt: new Date(),
        updatedAt: new Date(),
        type: 'text',
        page: {
          create: {
            handle: 'staff',
            displayName: '@staff',
            owner: {
              create: {
                id: randomUUID(),
              },
            },
            moderators: {
              create: createMockUser({
                id: session.user.id,
              }),
            },
          },
        },
      },
    });

    // @ts-expect-error - Mocking useQuery
    jest.mocked(api.post.getPostDetails.useQuery).mockImplementation(() => ({
      isLoading: false,
      data: post,
    }));

    const { container } = render(<PostId postId={post.id} />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('div')).toHaveTextContent('__TEXT_POST__MOCK__');
  });
});
