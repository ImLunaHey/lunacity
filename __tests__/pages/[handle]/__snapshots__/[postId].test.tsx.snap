// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostId renders a loading state 1`] = `
<div>
  <div
    class="nextui-c-bOoPrE nextui-c-bOoPrE-byitVA-color-currentColor nextui-c-bOoPrE-fQoxSm-textColor-default"
  >
    <span
      aria-label="Loading"
      class="nextui-c-chTLhx nextui-c-chTLhx-eWttbs-size-sm nextui-c-chTLhx-heRREU-type-points-opacity nextui-c-chTLhx-gzjEDO-cv nextui-c-chTLhx-iPJLV-css nextui-loading nextui-loading-points-opacity"
    >
      <i
        class="_1"
      />
      <i
        class="_2"
      />
      <i
        class="_3"
      />
    </span>
  </div>
</div>
`;

exports[`PostId renders a post with the provided ID 1`] = `
<div>
  <div>
    __TEXT_POST__MOCK__
  </div>
</div>
`;

exports[`PostId throws an error when no postId is provided 1`] = `
<div>
  <div
    style="font-family: -apple-system, BlinkMacSystemFont, Roboto, "Segoe UI", "Fira Sans", Avenir, "Helvetica Neue", "Lucida Grande", sans-serif; height: 100vh; text-align: center; display: flex; flex-direction: column; align-items: center; justify-content: center;"
  >
    <div>
      <style>
        
                body { margin: 0; color: #000; background: #fff; }
                .next-error-h1 {
                  border-right: 1px solid rgba(0, 0, 0, .3);
                }

                @media (prefers-color-scheme: dark) {
                  body { color: #fff; background: #000; }
                  .next-error-h1 {
                    border-right: 1px solid rgba(255, 255, 255, .3);
                  }
                }
      </style>
      <h1
        class="next-error-h1"
        style="display: inline-block; margin: 0px 20px 0px 0px; padding: 0px 23px 0px 0px; font-size: 24px; font-weight: 500; vertical-align: top; line-height: 49px;"
      >
        404
      </h1>
      <div
        style="display: inline-block; text-align: left; line-height: 49px; height: 49px; vertical-align: middle;"
      >
        <h2
          style="font-size: 14px; font-weight: normal; line-height: 49px; margin: 0px; padding: 0px;"
        >
          No post exists with this ID
          .
        </h2>
      </div>
    </div>
  </div>
</div>
`;
